runtime/shared_common/CompositeCache.cpp

```diff
@@ -6680,3 +6791,237 @@ SH_CompositeCacheImpl::hasReadMutex(J9VMThread* currentThread) const
+		goto done;
+	}
+	memcpy(headerCopy, _theca, headerSize);
+	
+	headerCopy->totalBytes += (freeDebugSpaceToChange + freeSpaceToChange);
+	headerCopy->updateSRP += (freeDebugSpaceToChange + freeSpaceToChange);
+	headerCopy->segmentSRP += freeDebugSpaceToChange;
+	headerCopy->crcValid = 0; /* segment will be changed, so set crcValid 0 */
+	/* A flag can be added in extraFlags to indicate the cache has been resized for debugging purpose. */
+	headerCopy->debugRegionSize += freeDebugSpaceToChange;
+	headerCopy->localVariableTableNextSRP += freeDebugSpaceToChange;
+	headerCopy->sharedStringHead += freeDebugSpaceToChange;
+	headerCopy->sharedStringTail += freeDebugSpaceToChange;
+	/* 
+	 * Unset minAOT/maxAOT/minJIT/maxJIT/softMaxBytes. As the cache is being resized, no need to care about such limits set on the orignal cache.
+	 * In addition, the original values might not be feasible anymore.
+	 * (e.g. minAOT + minJIT could become greater than new totalBytes, softMaxBytes could become smaller than used bytes in the new cache).
+	 */
+	headerCopy->minAOT = -1;
+	headerCopy->maxAOT = -1;
+	headerCopy->minJIT = -1;
+	headerCopy->maxJIT = -1;
+	headerCopy->softMaxBytes = (U_32)-1;
+	if (headerCopy->cacheFullFlags > 0) {
+		if (freeSpaceToChange > 0) {
+			/* Unset cacheFullFlags if the cache size is being increased. */
+			headerCopy->cacheFullFlags = 0;
+		}
+	}
+
+	if (headerSize != j9file_write(fd, (void*)headerCopy, headerSize)) {
+		I_32 errorno = j9error_last_error_number();
+		const char * errormsg = j9error_last_error_message();
+
+		CC_ERR_TRACE1(J9NLS_SHRC_PORT_ERROR_NUMBER, errorno);
+		Trc_SHR_Assert_True(errormsg != NULL);
+		CC_ERR_TRACE1(J9NLS_SHRC_PORT_ERROR_MESSAGE, errormsg);
+		goto done;
+	}
+	rc = true;
+done:
+	if (NULL != headerCopy) {
+		j9mem_free_memory(headerCopy);
+	}
+	return rc;
+}
+
+bool
+SH_CompositeCacheImpl::fixAndWriteDebugAreaAndClassSegment(J9VMThread* currentThread, IDATA fd, I_32 freeDebugSpaceToChange)
+{
+	PORT_ACCESS_FROM_VMC(currentThread);
+	I_32 debugAreaSize = (I_32)getDebugBytes();
+	I_32 LNTSize = (I_32)getLineNumberTableBytes();
+	BlockPtr LNTCopy = NULL;
+	BlockPtr debugStartAddr = (BlockPtr)getClassDebugDataStartAddress();
+	I_32 readWriteBytes = (I_32)getReadWriteBytes();
+
+	if (debugAreaSize >0) {
+		j9file_seek(fd, debugAreaSize + freeDebugSpaceToChange + readWriteBytes, EsSeekCur);
+		if (LNTSize > 0) {
+			LNTCopy = (BlockPtr)j9mem_allocate_memory(LNTSize, J9MEM_CATEGORY_VM);
+			if (NULL == LNTCopy) {
+				CC_ERR_TRACE(J9NLS_SHRC_OSCACHE_ALLOC_FAILED);
+				return false;
+			}
+			memcpy(LNTCopy, debugStartAddr, LNTSize);
+		}
+	}
+
+	BlockPtr walk = (BlockPtr)getBaseAddress();
+	BlockPtr prev = NULL;
+	BlockPtr endOfROMSegment = (BlockPtr)getSegmentAllocPtr();
+	I_32 romSegBytesWritten = 0;
+	while (walk < endOfROMSegment) {
+		U_32 romSize = ((J9ROMClass*)walk)->romSize;
+		J9ROMClass *romClassCopy = (J9ROMClass *)j9mem_allocate_memory(romSize, J9MEM_CATEGORY_VM);
```

Comment:
This should allocate a reasonable sized chunk once before walking (128K), and then reuse it for each class as long as it's big enough.

Reply:
Fixed.